packages/playwright-core/src/server/protocolError.ts

```diff
@@ -14,15 +14,33 @@
  * limitations under the License.
  */
 
+import { rewriteErrorMessage } from '../utils/stackTrace';
+
 export class ProtocolError extends Error {
-  sessionClosed: boolean;
+  type: 'error' | 'closed' | 'crashed';
+  method: string | undefined;
+  logs: string | undefined;
+
+  constructor(type: 'error' | 'closed' | 'crashed', method?: string, logs?: string) {
+    super();
+    this.type = type;
+    this.method = method;
+    this.logs = logs;
+  }
 
-  constructor(sessionClosed: boolean, message?: string) {
-    super(message);
-    this.sessionClosed = sessionClosed || false;
+  setMessage(message: string) {
+    rewriteErrorMessage(this, `Protocol error (${this.method}): ${message}`);
   }
+
+  browserLogMessage() {
+    return this.logs ? '\nBrowser logs:\n' + this.logs : '';
```

Comment:
Given that we use this for xserver and stuff, perhaps rename it to just `logs` and move the `'Browser logs'` prefix to the constructor call site?

Reply:
Let me follow up